### PR TITLE
Update attentivefp.py

### PR DESCRIPTION
**Pull request by xiaohongniua**
apply_edges1 and apply_edges2 fixed

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by mufeili** [comment]
Hi, thank you for the PR. Do you have a reference for the fix? (E.g. according to particular equations in the paper or particular lines of code in the original implementation)

**Comment by xiaohongniua** [comment]
> Hi, thank you for the PR. Do you have a reference for the fix? (E.g. according to particular equations in the paper or particular lines of code in the original implementation)

def forward(self, atom_list, bond_list, atom_degree_list, bond_degree_list, atom_mask):
        atom_mask = atom_mask.unsqueeze(2)
        batch_size,mol_length,num_atom_feat = atom_list.size()
        atom_feature = F.leaky_relu(self.atom_fc(atom_list))
        bond_neighbor = [bond_list[i][bond_degree_list[i]] for i in range(batch_size)]
        bond_neighbor = torch.stack(bond_neighbor, dim=0)
        atom_neighbor = [atom_list[i][atom_degree_list[i]] for i in range(batch_size)]
        atom_neighbor = torch.stack(atom_neighbor, dim=0)
        #then catenate them
        neighbor_feature = torch.cat([atom_neighbor, bond_neighbor],dim=-1)**#Concate neighbor atoms(mighe be dst node in dgl) and neighbor bonds** 
        neighbor_feature = F.leaky_relu(self.neighbor_fc(neighbor_feature))

        # generate mask to eliminate the influence of blank atoms
        attend_mask = atom_degree_list.clone()
        attend_mask[attend_mask != mol_length-1] = 1
        attend_mask[attend_mask == mol_length-1] = 0
        attend_mask = attend_mask.type(torch.cuda.FloatTensor).unsqueeze(-1)

        softmax_mask = atom_degree_list.clone()
        softmax_mask[softmax_mask != mol_length-1] = 0
        softmax_mask[softmax_mask == mol_length-1] = -9e8 # make the softmax value extremly small
        softmax_mask = softmax_mask.type(torch.cuda.FloatTensor).unsqueeze(-1)

        batch_size, mol_length, max_neighbor_num, fingerprint_dim = neighbor_feature.shape
        atom_feature_expand = atom_feature.unsqueeze(-2).expand(batch_size, mol_length, max_neighbor_num, fingerprint_dim)
        feature_attention = torch.cat([atom_feature_expand, neighbor_feature],dim=-1)

        attention_weight = self.attention[0](self.dropout(feature_attention))
#             print(attention_weight)
        attention_weight = attention_weight + softmax_mask
        attention_weight = F.softmax(attention_weight,-2)
#             print(attention_weight)
        attention_weight = attention_weight * attend_mask
#         print(attention_weight)
        neighbor_feature_transform = self.attend[0](self.dropout(neighbor_feature))
#             print(features_neighbor_transform.shape)
        attention_out = torch.sum(torch.mul(attention_weight,neighbor_feature_transform),-2)
#             print(attention_out.shape)
        attention_out = F.elu(attention_out)
        attention_out_reshape = attention_out.view(batch_size*mol_length, fingerprint_dim)
        atom_feature_reshape = atom_feature.view(batch_size*mol_length, fingerprint_dim)
        atom_feature_reshape = self.GRUCell[0](attention_out_reshape, atom_feature_reshape)
        atom_feature = atom_feature_reshape.view(batch_size, mol_length, fingerprint_dim)

        #do nonlinearity
        atom_feature = F.relu(atom_feature)

        for d in range(self.radius-1):
            # bonds_indexed = [bond_list[i][torch.cuda.LongTensor(bond_degree_list)[i]] for i in range(batch_size)]
            neighbor_feature = [atom_feature[i][atom_degree_list[i]] for i in range(batch_size)]

**Comment by xiaohongniua** [comment]
> Hi, thank you for the PR. Do you have a reference for the fix? (E.g. according to particular equations in the paper or particular lines of code in the original implementation)

`
    def forward(self, atom_list, bond_list, atom_degree_list, bond_degree_list, atom_mask):
        atom_mask = atom_mask.unsqueeze(2)
        batch_size,mol_length,num_atom_feat = atom_list.size()
        atom_feature = F.leaky_relu(self.atom_fc(atom_list))
        bond_neighbor = [bond_list[i][bond_degree_list[i]] for i in range(batch_size)]
        bond_neighbor = torch.stack(bond_neighbor, dim=0)
        atom_neighbor = [atom_list[i][atom_degree_list[i]] for i in range(batch_size)]
        atom_neighbor = torch.stack(atom_neighbor, dim=0)
        #then catenate them
        #**Concate neighbor atoms features and neighbor bond features to caculate attention weights**
        #**neighbor atoms mighe be dst nodes in dgl**
        neighbor_feature = torch.cat([atom_neighbor, bond_neighbor],dim=-1)
        neighbor_feature = F.leaky_relu(self.neighbor_fc(neighbor_feature))

        # generate mask to eliminate the influence of blank atoms
        attend_mask = atom_degree_list.clone()
        attend_mask[attend_mask != mol_length-1] = 1
        attend_mask[attend_mask == mol_length-1] = 0
        attend_mask = attend_mask.type(torch.cuda.FloatTensor).unsqueeze(-1)

        softmax_mask = atom_degree_list.clone()
        softmax_mask[softmax_mask != mol_length-1] = 0
        softmax_mask[softmax_mask == mol_length-1] = -9e8 # make the softmax value extremly small
        softmax_mask = softmax_mask.type(torch.cuda.FloatTensor).unsqueeze(-1)

        batch_size, mol_length, max_neighbor_num, fingerprint_dim = neighbor_feature.shape
        atom_feature_expand = atom_feature.unsqueeze(-2).expand(batch_size, mol_length, max_neighbor_num, fingerprint_dim)
        feature_attention = torch.cat([atom_feature_expand, neighbor_feature],dim=-1)

        attention_weight = self.attention[0](self.dropout(feature_attention))
#             print(attention_weight)
        attention_weight = attention_weight + softmax_mask
        attention_weight = F.softmax(attention_weight,-2)
#             print(attention_weight)
        attention_weight = attention_weight * attend_mask
#         print(attention_weight)
        neighbor_feature_transform = self.attend[0](self.dropout(neighbor_feature))
#             print(features_neighbor_transform.shape)
        attention_out = torch.sum(torch.mul(attention_weight,neighbor_feature_transform),-2)
#             print(attention_out.shape)
        attention_out = F.elu(attention_out)
        attention_out_reshape = attention_out.view(batch_size*mol_length, fingerprint_dim)
        atom_feature_reshape = atom_feature.view(batch_size*mol_length, fingerprint_dim)
        atom_feature_reshape = self.GRUCell[0](attention_out_reshape, atom_feature_reshape)
        atom_feature = atom_feature_reshape.view(batch_size, mol_length, fingerprint_dim)

        #do nonlinearity
        atom_feature = F.relu(atom_feature)

        for d in range(self.radius-1):
            # bonds_indexed = [bond_list[i][torch.cuda.LongTensor(bond_degree_list)[i]] for i in range(batch_size)]
            neighbor_feature = [atom_feature[i][atom_degree_list[i]] for i in range(batch_size)]
`

**Comment by xiaohongniua** [comment]
> Hi, thank you for the PR. Do you have a reference for the fix? (E.g. according to particular equations in the paper or particular lines of code in the original implementation)

Sorry, I might have mistoken the "neighbor nodes" in dgl, 
I will close this request!
Sorry.

**Comment by mufeili** [comment]
> > Hi, thank you for the PR. Do you have a reference for the fix? (E.g. according to particular equations in the paper or particular lines of code in the original implementation)
> 
> Sorry, I might have mistoken the "neighbor nodes" in dgl,
> I will close this request!
> Sorry.

No worries. Still thank you for the PR.
